src/polly/src/event_manager.rs

```diff
@@ -52,75 +52,60 @@ pub trait Subscriber {
     ///
     /// # Arguments
     /// * event - the available `EpollEvent` ready for processing
-    /// * event_manager - Reference to the `EventManager` that gives the implementor
-    ///                   the possibility to directly call the required update operations.
-    ///                   The only functions safe to call on this `EventManager` reference
-    ///                   are `register`, `unregister` and `modify` which correspond to
-    ///                   the `libc::epoll_ctl` operations.
-    fn process(&mut self, event: EpollEvent, event_manager: &mut EventManager);
+    /// * event_manager - Mutable reference to the `EventManager`. It allows
+    ///                   updating the events that the subscriber is interested
+    ///                   in.
+    fn process(&mut self, event: EpollEvent, event_manager: &mut dyn EventManager);
 
     /// Returns a list of `EpollEvent` that this subscriber is interested in.
     fn interest_list(&self) -> Vec<EpollEvent>;
 }
 
+pub trait EventManager {
+    /// Registers a new event.
+    ///
+    /// # Arguments
+    /// * pollable - target pollable object.
+    /// * epoll_event - configuration of the event being registered.
+    /// * subscriber - subscriber responsible for processing the event.
+    fn register(
+        &mut self,
+        pollable: Pollable,
+        epoll_event: EpollEvent,
+        subscriber: Arc<Mutex<dyn Subscriber>>,
+    ) -> Result<()>;
+
+    /// Removes `pollable` from the interest list.
+    fn unregister(&mut self, pollable: Pollable) -> Result<()>;
+
+    /// Updates the event configuration corresponding to `pollable`.
+    ///
+    /// # Arguments
+    /// * pollable - target pollable object.
+    /// * epoll_event - new event configuration.
+    fn modify(&mut self, pollable: Pollable, epoll_event: EpollEvent) -> Result<()>;
```

Comment:
The trait is to be generic, but ends up being very specific when you mention epoll. It kind of asks for poll event abstraction or at least renaming 😄

Reply:
The purpose of the trait for now is to fix the other problem generated by passing the event manager to the `process` function (i.e. you could call `run` and be in an endless loop). As you mentioned, the trait and implementation needs further enhancements to make it less epoll specific.

Reply:
Should we open an issue for removing this specificity?

Reply:
I was thinking about consuming this from rust-vmm and fix these problems directly there so we don't have to review PRs in Firecracker and rust-vmm. What do you think?